hazelcast/src/main/java/com/hazelcast/map/impl/operation/PutAllOperation.java

```diff
@@ -61,59 +64,72 @@ public PutAllOperation(String name, MapEntries mapEntries) {
+    }
+
+    private boolean hasBackups() {
+        return (mapContainer.getTotalBackupCount() > 0);
+    }
+
     private void put(Map.Entry<Data, Data> entry) {
         Data dataKey = entry.getKey();
         Data dataValue = entry.getValue();
 
-        Object oldValue = null;
-        if (hasMapListener) {
-            oldValue = recordStore.put(dataKey, dataValue, DEFAULT_TTL);
-        } else {
-            // By using `recordStore.set`, we get-rid-of one extra map-store access.
-            // Because `recordStore.put` tries to find previous value from map-store, in order to pass it EntryEvent.
-            // If loading from map-store is expensive, this can lead serious performance degradation.
-            // To prevent this potential problem, when there is no map-listener exists, don't use `recordStore.put`.
-            recordStore.set(dataKey, dataValue, DEFAULT_TTL);
-        }
-
+        Object oldValue = putToRecordStore(dataKey, dataValue);
         dataValue = getValueOrPostProcessedValue(dataKey, dataValue);
         mapServiceContext.interceptAfterPut(name, dataValue);
 
         if (hasMapListener) {
-            EntryEventType eventType = oldValue == null ? ADDED : UPDATED;
+            EntryEventType eventType = (oldValue == null ? ADDED : UPDATED);
             mapEventPublisher.publishEvent(getCallerAddress(), name, eventType, dataKey, oldValue, dataValue);
         }
 
-        Record record = recordStore.getRecord(dataKey);
-
-        if (shouldWanReplicate) {
+        Record record = (hasWanReplication || hasBackups) ? recordStore.getRecord(dataKey) : null;
+        if (hasWanReplication) {
             EntryView entryView = createSimpleEntryView(dataKey, dataValue, record);
             mapEventPublisher.publishWanReplicationUpdate(name, entryView);
         }
-        backupEntries.add(entry);
-        RecordInfo replicationInfo = buildRecordInfo(record);
-        backupRecordInfos.add(replicationInfo);
-        evict();
+        if (hasBackups) {
+            backupEntries.add(entry);
+            RecordInfo replicationInfo = buildRecordInfo(record);
+            backupRecordInfos.add(replicationInfo);
+        }
 
-        addInvalidation(dataKey);
+        evict();
+        if (hasInvalidation) {
+            invalidationKeys.add(dataKey);
+        }
     }
 
-    private void addInvalidation(Data dataKey) {
-        if (mapContainer.isInvalidationEnabled()) {
-            if (invalidationKeys == null) {
-                invalidationKeys = new ArrayList<Data>(mapEntries.size());
-            }
-            invalidationKeys.add(dataKey);
+    /**
+     * The method recordStore.put() tries to fetch the old value from the MapStore,
+     * which can lead to a serious performance degradation if loading from MapStore is expensive.
+     * We prevent this by calling recordStore.set() if no map listeners are registered.
+      */
+    private Object putToRecordStore(Data dataKey, Data dataValue) {
```

Comment:
I had to reduce the NPath complexity of `put()`, so I decided to pull out this method, since the code block already needed a comment to explain why there is put() and set() being used. So this is a nice JavaDoc now :)